endpoints/openrtb2/auction.go

```diff
@@ -332,26 +339,43 @@ func (deps *endpointDeps) validateImpExt(ext openrtb.RawJSON, impIndex int) erro
 	}
 
 	for bidder, ext := range bidderExts {
-		bidderName, isValid := openrtb_ext.BidderMap[bidder]
-		if isValid {
-			if err := deps.paramsValidator.Validate(bidderName, ext); err != nil {
-				return fmt.Errorf("request.imp[%d].ext.%s failed validation.\n%v", impIndex, bidder, err)
+		if bidder != "prebid" {
+			coreBidder := bidder
+			if tmp, isAlias := aliases[bidder]; isAlias {
+				coreBidder = tmp
+			}
+			if bidderName, isValid := openrtb_ext.BidderMap[coreBidder]; isValid {
+				if err := deps.paramsValidator.Validate(bidderName, ext); err != nil {
+					return fmt.Errorf("request.imp[%d].ext.%s failed validation.\n%v", impIndex, coreBidder, err)
+				}
+			} else {
+				return fmt.Errorf("request.imp[%d].ext contains unknown bidder: %s. Did you forget an alias in request.ext.prebid.aliases?", impIndex, bidder)
 			}
-		} else if bidder != "prebid" {
-			return fmt.Errorf("request.imp[%d].ext contains unknown bidder: %s", impIndex, bidder)
 		}
 	}
 
 	return nil
 }
 
-func (deps *endpointDeps) validateBidRequestExt(ext openrtb.RawJSON) error {
+func (deps *endpointDeps) parseBidExt(ext openrtb.RawJSON) (*openrtb_ext.ExtRequest, error) {
 	if len(ext) < 1 {
-		return nil
+		return nil, nil
 	}
 	var tmpExt openrtb_ext.ExtRequest
 	if err := json.Unmarshal(ext, &tmpExt); err != nil {
-		return fmt.Errorf("request.ext is invalid: %v", err)
+		return nil, fmt.Errorf("request.ext is invalid: %v", err)
+	}
+	return &tmpExt, nil
+}
+
+func (deps *endpointDeps) validateAliases(aliases map[string]string) error {
+	for thisAlias, coreBidder := range aliases {
+		if _, isCoreBidder := openrtb_ext.BidderMap[coreBidder]; !isCoreBidder {
+			return fmt.Errorf("request.ext.prebid.aliases.%s refers to unknown bidder: %s", thisAlias, coreBidder)
+		}
+		if thisAlias == coreBidder {
+			return fmt.Errorf("request.ext.prebid.aliases.%s defines a no-op alias. Choose a different alias, or remove this entry.", thisAlias)
```

Comment:
Should this necessarily return an error?

Reply:
definitely debatable... I'm not 100% sure.

My thoughts were: "If they sent this, we would ignore it. If we ignore it, then they'll get better performance (smaller I/O) if they don't send it. Let's encourage that."

I sorta doubt anyone will care either way, though.

Reply:
It is debatable. I was thinking there is one case that we would probably want to return an error, but this wouldn't catch it. Do we want to flag when someone aliases one core bidder to another? Like appnexus=rubicon? I don't really see a use case to do this, and if someone does it is quite likely that it will just cause confusion. Perhaps we want to check for this case as well?

We could use a method to pass back warnings, so we can flag curious things like this without invalidation the entire bid request for the bidder. Developers could then see the message when they are trying to debug why the auction isn't happening as they expected. But that is probably out of scope for this PR.

Reply:
Yeah, the `appnexus=rubicon` one is a good case. I definitely did consider making that an error... but that would actually make every new Bidder a breaking change. Today, publishers _could_ alias `foo=appnexus`, and tomorrow, someone could submit a new Bidder with the name `foo`. It's very unlikely, but.... If we make that an error, then those calls would start breaking.

Instead I made `appnexus=rubicon` legal, and defined the API so that aliases take precedence over "core" bidders. There's a section about it in [the aliases docs](https://github.com/prebid/prebid-server/blob/master/docs/endpoints/openrtb2/auction.md#bidder-aliases)

> If an alias overlaps with a core Bidder's name, then the alias will take precedence. This prevents breaking API changes as new Bidders are added to the project.

Reply:
I've been toying with the idea of "warn, but execute the auction anyway" too... but I lean moderately against it.

Devs are more likely to miss or ignore 2xx responses with an error message stuffed into them than they are 4xx's. If the error message is good, it takes like 30 seconds to fix.

In some cases it's obvious what the user "meant" and we can run an appropriate auction, but in other cases it's not. It's like a deer in a thicket; you can walk in easily, but you're likely to get your antlers tangled, and then it'll be impossible to escape. And if they miss the warnings, and the auction doesn't behave _exactly_ like they expect, they'll log issues about it... which we'll then have to debug.

I'd just rather not go down that road.